libraries/botbuilder/tests/channelServiceRoutes.test.js

```diff
@@ -0,0 +1,776 @@
+                        done();
+                    };
+
+                    const service = sinon.createStubInstance(ChannelServiceHandler);
+                    service.handleUploadAttachment = sinon.stub().rejects(resourceResponse);
+
+                    const channel = new ChannelServiceRoutes(service);
+                    channel.processUploadAttachment(req, res);
+                } catch (error) {
+                    done(error);
+                }
+            });
+
+            it('should throw a readBody error', done => {
+                try {
+                    const resourceResponse = { error: 'readBody error' };
+                    const res = new MockResponse({
+                        statusCode: 200,
+                        body: resourceResponse
+                    }, done);
+
+                    ChannelServiceRoutes.handleError = (err) => {
+                        assert.deepStrictEqual(err, resourceResponse, `expected: ${ JSON.stringify(resourceResponse) }. received: ${ JSON.stringify(err) }`);
+                        done();
+                    };
+
+                    const service = sinon.createStubInstance(ChannelServiceHandler);
+
+                    readBodyStub = sinon.stub(ChannelServiceRoutes, 'readBody');
+                    readBodyStub.rejects(resourceResponse);
+
+                    const channel = new ChannelServiceRoutes(service);
+                    channel.processUploadAttachment(req, res);
+                } catch (error) {
+                    done(error);
+                }
+            });
+        });
+
+        describe('readActivity()', async () => {
+            it('should throw with invalid body', async () => {
+                const req = sinon.mock(WebRequest);
+                req.body = {};
+
+                ChannelServiceRoutes.readActivity(req).catch(err => {
+                    assert.strictEqual(err.statusCode, StatusCodes.BAD_REQUEST);
+                });
+            });
+
+            it('should return activity', async () => {
+                const req = sinon.mock(WebRequest);
+                req.body = {
+                    type: 'testactivity',
+                    timestamp: Date.now(),
+                    expiration: '2200-01-01',
+                    localTimeStamp: Date.now(),
+                };
+
+                ChannelServiceRoutes.readActivity(req).then((activity) => {
+                    assert.strictEqual(activity.type, 'testactivity');
+                });
+            });
+        });
+
+        describe('readBody()', async () => {
+            it('should return request.body from "on" events', async () => {
+                const source = { test: true };
+
+                const request = {
+                    ...req,
+                    body: null
+                };
+
+                setTimeout(() => {
```

Comment:
Can this be rewritten to avoid using a setTimeout?  Or, please add some comments explaining why it is necessary for this test.